[CLEANUP] Refacto la récupération des réponses sur les QROCM (PIX-650).

jonasgrilleres opened the pull request:
## :unicorn: Problème
Un FIXME traîne depuis un moment sur la façon de récupérer les réponses depuis les input fields sur une page d'épreuve. 

## :robot: Solution
Utiliser le two-way binding pour stocker le contenu des champs réponse dans une variable, plutôt que de lire depuis le DOM.

## :rainbow: Remarques
On supprime le helper `property-of` qui n'est plus utilisé.

## :100: Pour tester
Aller sur cette épreuve : **rec25lceSer8ri5TH** : 
- Essayer de valider en ne remplissant pas tous les champs. On doit avoir une erreur.
- Après avoir répondu à l'épreuve, revenir dessus et constater que les réponses données sont correctement pré-remplies. 
- Voir dans le checkpoint les réponses données.

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr1711.review.pix.fr
- Orga: https://orga-pr1711.review.pix.fr
- Certif: https://certif-pr1711.review.pix.fr
- Admin: https://admin-pr1711.review.pix.fr
- API: https://api-pr1711.review.pix.fr/api/

 Please check it out!

HEYGUL commented on mon-pix/app/templates/components/qrocm-proposal.hbs:
```diff
@@ -18,7 +18,7 @@
                 disabled={{answer}}
                 aria-label={{block.ariaLabel}}
-                onkeydown={{action 'onInputChange'}}>
+                onkeyup={{action 'onInputChange'}}>
```
on peut adopter la syntaxe 
```
{{on 'keyup' this.onInputChange}}
```

HEYGUL commented on mon-pix/app/components/challenge-item-qrocm.js:
```diff
@@ -41,6 +38,11 @@ class ChallengeItemQrocm extends ChallengeItemGeneric {
 
   @action
   answerChanged() {
+    const inputName = event.currentTarget.name;
```
Il faudrait passer l'event en argument non ?

HEYGUL commented on mon-pix/app/components/challenge-item-qrocm.js:
```diff
@@ -27,12 +29,7 @@ class ChallengeItemQrocm extends ChallengeItemGeneric {
   // This is not "the Ember way", however it makes code easier to read,
   // and moreover, is a much more robust solution when you need to test it properly.
   _getRawAnswerValue() {
```
Supprimer le commentaire obsolète.

HEYGUL commented on mon-pix/app/components/challenge-item-qrocm.js:
```diff
@@ -41,6 +38,11 @@ class ChallengeItemQrocm extends ChallengeItemGeneric {
+    const inputName = event.currentTarget.name;
+    const inputValue = event.currentTarget.value;
+
+    this.answerValues = {...this.answerValue, [inputName]: inputValue };
```
Pourquoi `[inputName]` et non `` `${inputName}` `` ?

celineung replied:
```suggestion
    this.answerValues = { ...this.answerValues, [inputName]: inputValue };
```

HEYGUL commented on mon-pix/app/components/challenge-item-qrocm.js:
```diff
@@ -27,12 +29,7 @@ class ChallengeItemQrocm extends ChallengeItemGeneric {
-      result[element.getAttribute('name')] = element.value;
-    });
-    return result;
+    return this.answerValues;
```
Du coup, la fonction `_getRawAnswerValue` peut être remplacée par `this.answerValues` directement.

HEYGUL commented:
Je viens de regarder plus en détails, notamment les tests qui ne passent pas.
Plusieurs soucis : 
 - On a une propriété `answerValue` et une propriété `answerValues`, cela risque de prêter à confusion. On peut renommer `answerValues` par `userInputValues`.
 - On réagit désormais au `keyup` : il faut donc modifier les tests en conséquence.
 - On ne boucle pas sur tous les éléments `input`, du coup si on ne répond qu'à un seul champ, on pense que l'utilisateur a bien rempli tous les champs. Il faudrait une initialisation de `answersValues` avec tous les champs du QROCM.

HEYGUL commented on mon-pix/app/components/challenge-item-qrocm.js:
```diff
@@ -10,6 +10,8 @@ import jsyaml from 'js-yaml';
 class ChallengeItemQrocm extends ChallengeItemGeneric {
   @service intl;
 
+  answerValues = [];
```
Ce n'est pas un tableau mais un objet, on devrait le changer par 
```suggestion
  answerValues = {};
```
De plus, sans initialisation avec tous les champs, on ne peut détecter un remplissage partiel.
Il faudrait qu'à l'init, on est un objet de ce genre :
```
answerValues = { field1: '', field2: '', field3: '' };
```

jonasgrilleres committed:
Two-way bind answer fields with the controller

Instead of reading the values from the DOM, we use the two-way binding offered by the template to store the answers given by the user into a map.
More details here: [Binding Dynamic Attributes](https://guides.emberjs.com/v3.20.0/components/built-in-components/#toc_binding-dynamic-attribute).

jonasgrilleres committed:
Delete unused custom helper

HEYGUL commented:
Testé fonctionnellement 👍

MelanieMEB commented on mon-pix/app/components/challenge-item-qrocm.js:
```diff
@@ -5,13 +5,29 @@ import classic from 'ember-classic-decorator';
+  answersValue = {};
+
+  didReceiveAttrs() {
+    const proposals = proposalsAsBlocks(this.challenge.proposals);
+    proposals.forEach(({ input }) => {
+      if (input) {
+        this.answersValue[input] = '';
+      }
+    });
+
```
Je propose de sortir ça dans une fonction pour préciser ce qui est fait
